### PR TITLE
Hw9

### PR DESCRIPTION
**Pull request by Abbath90**

### CODE REVIEW

File: homework9/task3/file_counter.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+from pathlib import Path
+from typing import Callable, Optional, TextIO
+
+
+def tokenizer_generator(file_handler: TextIO, tok: Optional[Callable]):
+    buffer = ""
+    char = " "
+    while char:
+        char = file_handler.read(1)
+        if not char:
```
Comment (AlekseiAfanasev) [inline homework9/task3/file_counter.py:22]: if we use `if not char` here, then `while True` loop can be used

File: homework9/task3/file_counter.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+#>>> universal_file_counter(test_dir, "txt", str.split)
+6
+"""
+import glob
+import os
+from pathlib import Path
+from typing import Callable, Optional, TextIO
+
+
+def tokenizer_generator(file_handler: TextIO, tok: Optional[Callable]):
```
Comment (AlekseiAfanasev) [inline homework9/task3/file_counter.py:17]: retiurn typings

File: homework9/task3/file_counter.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+) -> int:
+    counter = 0
+    for file in Path(dir_path).glob("*." + file_extension):
+        with open(file) as f:
+            if tokenizer is not None:
+                for token in tokenizer_generator(f, tokenizer):
+                    counter += token
+            else:
+                for _ in f:
+                    print(_)
```
Comment (AlekseiAfanasev) [inline homework9/task3/file_counter.py:45]: redundant?

File: homework9/task3/file_counter.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+from typing import Callable, Optional, TextIO
+
+
+def tokenizer_generator(file_handler: TextIO, tok: Optional[Callable]):
+    buffer = ""
+    char = " "
+    while char:
+        char = file_handler.read(1)
+        if not char:
+            yield 1  # buffer
```
Comment (AlekseiAfanasev) [inline homework9/task3/file_counter.py:23]: this will lead to an extra "1" in counter value, isn't it?

File: homework9/task3/file_counter.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    dir_path: Path, file_extension: str, tokenizer: Optional[Callable] = None
+) -> int:
+    counter = 0
+    for file in Path(dir_path).glob("*." + file_extension):
+        with open(file) as f:
+            if tokenizer is not None:
+                for token in tokenizer_generator(f, tokenizer):
+                    counter += token
+            else:
+                for _ in f:
```
Comment (AlekseiAfanasev) [inline homework9/task3/file_counter.py:44]: this loop can be simnplified to
`sum(1 for _ in f)`

File: homework9/tests/test_filecounter.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+def test_file_counter_without_tokenizer(tmpdir):
+    tmpdir.join("file1.txt").write(content1)
+    tmpdir.join("file2.txt").write(content2)
+    tmpdir.join("file3.file").write(content3)
+    actual_result = universal_file_counter(Path(tmpdir), "txt")
+    assert actual_result == 6
+
+
+def test_file_counter_with_tokenizer(tmpdir):
```
Comment (AlekseiAfanasev) [inline homework9/tests/test_filecounter.py:23]: missing test for case without tokenizer function (None)

File: homework9/task3/file_counter.py (hunk 4 above)
Comment (Abbath90) [reply homework9/task3/file_counter.py:23]: no, there is stay last token in buffer after file is ending

File: homework9/tests/test_filecounter.py (hunk 6 above)
Comment (Abbath90) [reply homework9/tests/test_filecounter.py:23]: test test_file_counter_without_tokenizer(...)  is for this case

File: homework9/task3/file_counter.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+#>>> universal_file_counter(test_dir, "txt", str.split)
+6
+"""
+import glob
+import os
+from pathlib import Path
+from typing import Callable, Optional, TextIO
+
+
+def tokenizer_generator(file_handler: TextIO, tok: Optional[Callable]) -> int:
```
Comment (AlekseiAfanasev) [inline homework9/task3/file_counter.py:17]: this is a generator function - so, it does not return int, it returns Generator
